server/src/main/java/org/elasticsearch/http/HttpRequest.java

```diff
@@ -11,51 +11,23 @@
 import org.elasticsearch.common.bytes.BytesReference;
 import org.elasticsearch.core.Nullable;
 import org.elasticsearch.rest.ChunkedRestResponseBody;
-import org.elasticsearch.rest.RestRequest;
 import org.elasticsearch.rest.RestStatus;
 
 import java.util.List;
-import java.util.Map;
 
 /**
  * A basic http request abstraction. Http modules needs to implement this interface to integrate with the
  * server package's rest handling.
  */
-public interface HttpRequest {
+public interface HttpRequest extends BasicHttpRequest {
```

Comment:
Is this a general preference of yours or it's just this particular case?

I think the two options are:
 * an interface extends another and adds more methods; in case the base interface is passed around and the code has some specialized logic it can use `instance of` checks to downcast
 * have a single rich interface and throw "not implemented"

Reply:
FWIW the body needed for auditing authn success (which, again, is one of the things that's not dealt with properly by this PR) so pending other discussions, we might not need a request interface with a request body....

Reply:
> don't use anonymous implementations of BasicHttpRequest instead use a concrete HttpRequestWithoutBody.

That's separate from the interface conversation, I think. There's going to be some light wrapping needed because the request classes that netty uses are not available to the Security plugin module.

Reply:
> Is this a general preference of yours or it's just this particular case?

I think general ? ... basically I think the interface should represent the concept and any sub interfaces should be specializations of that concept. (IMO) Here the concept is a http request and not having access to the body the a specialization.  I understand the throwing not implemented from many methods is also it's own issue, so this is more subjective / preference than any rule.  However, I do think as-is this has some readability / expectation issues. Maybe just changing the names would help ... as-is there no indication to what the difference between a HttpRequest and BasicHttpRequest means...java doc can help but good names > doc.  I think i still prefer what I suggested above ...but having something like FullHttpRequest extends PartialHttpRequest makes more readily apparent what to expect from the contract.

Reply:
> ... has some specialized logic it can use instance of checks to downcast

I would strongly advise against a design that necessitates instanceof checks to downcast